File: SQL Injection/sql_injection.c
#include <stdio.h>
#include <stdlib.h>
#include <sqlite3.h>
#include <string.h>

// --- Function Prototypes ---
void setup_database(sqlite3 *db);
void demonstrate_safe_fetching(sqlite3 *db);
void demonstrate_injection_prevention(sqlite3 *db);
void execute_sql(sqlite3 *db, const char *sql);
void check_error(int rc, char *err_msg, sqlite3 *db);

// --- Main Execution ---
int main(int argc, char* argv[]) {
    sqlite3 *db;
    char *err_msg = 0;
    const char *db_file = "production_c_example.db";

    // Open the database
    int rc = sqlite3_open(db_file, &db);

    if (rc != SQLITE_OK) {
        fprintf(stderr, "Cannot open database: %s\n", sqlite3_errmsg(db));
        sqlite3_close(db);
        return 1;
    }
    printf("Database connection opened to '%s'.\n", db_file);

    setup_database(db);

    printf("\n--- DEMONSTRATING SAFE AND CONVENIENT FETCHING ---\n");
    demonstrate_safe_fetching(db);

    printf("\n--- DEMONSTRATING SQL INJECTION PREVENTION ---\n");
    demonstrate_injection_prevention(db);

    // Close the database
    sqlite3_close(db);
    printf("Database connection to '%s' closed.\n", db_file);
    
    // Clean up the database file
    if (remove(db_file) == 0) {
        printf("Cleaned up %s.\n", db_file);
    } else {
        fprintf(stderr, "Error cleaning up %s.\n", db_file);
    }

    return 0;
}

// --- Helper Functions ---

void check_error(int rc, char *err_msg, sqlite3 *db) {
    if (rc != SQLITE_OK) {
        fprintf(stderr, "SQL error: %s\n", err_msg);
        sqlite3_free(err_msg);
        sqlite3_close(db);
        exit(1);
    }
}

void execute_sql(sqlite3 *db, const char *sql) {
    char *err_msg = 0;
    int rc = sqlite3_exec(db, sql, 0, 0, &err_msg);
    check_error(rc, err_msg, db);
}

void setup_database(sqlite3 *db) {
    printf("Setting up the 'users' table...\n");

    execute_sql(db, "DROP TABLE IF EXISTS users");
    execute_sql(db, 
        "CREATE TABLE users("
        "id INTEGER PRIMARY KEY, "
        "username TEXT NOT NULL UNIQUE, "
        "email TEXT);"
    );
    
    const char *users_to_add[][3] = {
        {"1", "alice", "alice@example.com"},
        {"2", "bob", "bob@example.com"},
        {"3", "charlie", "charlie@example.com"}
    };

    sqlite3_stmt *stmt;
    const char *sql = "INSERT INTO users (id, username, email) VALUES (?, ?, ?);";
    
    int rc = sqlite3_prepare_v2(db, sql, -1, &stmt, 0);
    if (rc != SQLITE_OK) {
        fprintf(stderr, "Failed to prepare statement: %s\n", sqlite3_errmsg(db));
        return;
    }

    for (int i = 0; i < 3; ++i) {
        sqlite3_bind_int(stmt, 1, atoi(users_to_add[i][0]));
        sqlite3_bind_text(stmt, 2, users_to_add[i][1], -1, SQLITE_STATIC);
        sqlite3_bind_text(stmt, 3, users_to_add[i][2], -1, SQLITE_STATIC);
        
        if (sqlite3_step(stmt) != SQLITE_DONE) {
            fprintf(stderr, "Failed to insert user %s: %s\n", users_to_add[i][1], sqlite3_errmsg(db));
        }
        sqlite3_reset(stmt);
    }

    sqlite3_finalize(stmt);
    printf("Database setup complete. Users inserted.\n");
}

void demonstrate_safe_fetching(sqlite3 *db) {
    sqlite3_stmt *stmt;
    const char *sql = "SELECT id, username FROM users ORDER BY id";

    int rc = sqlite3_prepare_v2(db, sql, -1, &stmt, 0);
    if (rc != SQLITE_OK) {
        fprintf(stderr, "Failed to prepare statement: %s\n", sqlite3_errmsg(db));
        return;
    }

    printf("[GOOD] Fetched all users:\n");
    while ((rc = sqlite3_step(stmt)) == SQLITE_ROW) {
        int id = sqlite3_column_int(stmt, 0);
        const unsigned char *username = sqlite3_column_text(stmt, 1);
        printf("  ID: %d, Username: %s\n", id, username);
    }
    sqlite3_finalize(stmt);

    // GOOD: Fetch one specific user
    int user_id = 2;
    sql = "SELECT * FROM users WHERE id = ?";
    rc = sqlite3_prepare_v2(db, sql, -1, &stmt, 0);
    if (rc != SQLITE_OK) {
        fprintf(stderr, "Failed to prepare statement: %s\n", sqlite3_errmsg(db));
        return;
    }
    sqlite3_bind_int(stmt, 1, user_id);

    if ((rc = sqlite3_step(stmt)) == SQLITE_ROW) {
        printf("[GOOD] Fetched user with ID %d: ID: %d, Username: %s, Email: %s\n", 
            user_id,
            sqlite3_column_int(stmt, 0),
            sqlite3_column_text(stmt, 1),
            sqlite3_column_text(stmt, 2)
        );
    }
    sqlite3_finalize(stmt);
}

void demonstrate_injection_prevention(sqlite3 *db) {
    sqlite3_stmt *stmt;
    const char *malicious_input = "2; DROP TABLE users; --";
    const char *sql = "SELECT * FROM users WHERE id = ?";

    int rc = sqlite3_prepare_v2(db, sql, -1, &stmt, 0);
    if (rc != SQLITE_OK) {
        fprintf(stderr, "Failed to prepare statement: %s\n", sqlite3_errmsg(db));
        return;
    }
    
    // The malicious input is bound as a single string parameter, not executed as SQL
    sqlite3_bind_text(stmt, 1, malicious_input, -1, SQLITE_STATIC);
    
    printf("[SAFE] Tried to fetch user with malicious ID. Result:\n");
    if ((rc = sqlite3_step(stmt)) == SQLITE_ROW) {
         printf("  Malicious query returned a row (it shouldn\'t have!): ID: %d, Name: %s\n", sqlite3_column_int(stmt, 0), sqlite3_column_text(stmt, 1));
    } else {
        printf("  No user found with ID '%s', as expected.\n", malicious_input);
    }
    sqlite3_finalize(stmt);

    // Verify that the users table still exists and is unharmed
    sql = "SELECT username FROM users";
    rc = sqlite3_prepare_v2(db, sql, -1, &stmt, 0);
    if (rc != SQLITE_OK) {
        fprintf(stderr, "Failed to prepare statement: %s\n", sqlite3_errmsg(db));
        return;
    }

    printf("[SAFE] Data is unharmed. Users remaining:\n");
    while ((rc = sqlite3_step(stmt)) == SQLITE_ROW) {
        const unsigned char *username = sqlite3_column_text(stmt, 0);
        printf("  - %s\n", username);
    }
    sqlite3_finalize(stmt);
}

File: SQL Injection/sql_injection.py
#!/usr/bin/env python3

import sqlite3
import re
import os
import logging
from typing import Any, Iterable, Optional, List

# --- Setup basic logging ---
logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(levelname)s - %(message)s')

class SafeDBWrapper:
    """
    A production-ready database wrapper that prevents SQL injection, provides
    context management for connections, and offers granular transaction control.

    This class enforces parameterized queries, the standard defense against
    SQL injection, by separating SQL logic from data.
    """

    def __init__(self, db_name: str):
        """Initializes the wrapper but does not open the connection yet."""
        self.db_name = db_name
        self.connection: Optional[sqlite3.Connection] = None
        self.cursor: Optional[sqlite3.Cursor] = None

    def __enter__(self):
        """
        Opens the database connection and returns the wrapper instance.
        Enables use of the `with` statement for automatic connection handling.
        """
        try:
            self.connection = sqlite3.connect(self.db_name)
            # Set row_factory to sqlite3.Row to get dict-like rows
            self.connection.row_factory = sqlite3.Row
            self.cursor = self.connection.cursor()
            logging.info(f"Database connection opened to '{self.db_name}'.")
            return self
        except sqlite3.Error as e:
            logging.error(f"Failed to connect to database: {e}")
            raise

    def __exit__(self, exc_type, exc_val, exc_tb):
        """
        Closes the database connection. If an exception occurred within the
        `with` block, it rolls back the transaction. Otherwise, it commits.
        """
        if self.connection:
            if exc_type:
                logging.warning(f"Exception occurred, rolling back transaction. Details: {exc_val}")
                self.connection.rollback()
            else:
                logging.info("Committing transaction.")
                self.connection.commit()
            self.connection.close()
            logging.info(f"Database connection to '{self.db_name}' closed.")

    def _pre_execution_check(self, query: str):
        """
        Performs a basic check for obviously unsafe query patterns.
        This is a secondary defense; the primary defense is parameterization.
        """
        # This regex is a simple heuristic to catch common mistakes.
        unsafe_pattern = re.compile(r"""\bWHERE\b\s+\w+\s*=\s*['"].*['"]""", re.IGNORECASE)
        if unsafe_pattern.search(query):
            raise ValueError(
                "Potential SQL Injection risk detected. Do not use string "
                "formatting or concatenation for values in a WHERE clause. "
                "Use '?' placeholders and the 'params' argument instead."
            )

    def execute(self, query: str, params: Iterable[Any] = ()):
        """
        Executes a SQL query safely using parameter substitution.
        Does NOT commit automatically, allowing for multiple operations in one transaction.

        Args:
            query (str): The SQL query with '?' placeholders.
            params (Iterable, optional): A tuple or list of values.

        Returns:
            sqlite3.Cursor: The cursor object for manual fetching if needed.
        """
        if not self.cursor:
            raise sqlite3.Error("Cursor is not available. Is the connection open?")
        self._pre_execution_check(query)
        logging.info(f"Executing query: {query}")
        try:
            return self.cursor.execute(query, params)
        except sqlite3.Error as e:
            logging.error(f"Database error on execute: {e}")
            raise

    def fetch_one(self) -> Optional[sqlite3.Row]:
        """Fetches a single result from the last executed query."""
        if not self.cursor:
            raise sqlite3.Error("Cursor is not available.")
        return self.cursor.fetchone()

    def fetch_all(self) -> List[sqlite3.Row]:
        """Fetches all results from the last executed query."""
        if not self.cursor:
            raise sqlite3.Error("Cursor is not available.")
        return self.cursor.fetchall()

    def execute_and_fetch_all(self, query: str, params: Iterable[Any] = ()) -> List[sqlite3.Row]:
        """A convenience method to execute a query and fetch all results."""
        self.execute(query, params)
        return self.fetch_all()

    def execute_and_fetch_one(self, query: str, params: Iterable[Any] = ()) -> Optional[sqlite3.Row]:
        """A convenience method to execute a query and fetch a single result."""
        self.execute(query, params)
        return self.fetch_one()


# --- New and Improved Usage Example ---

def setup_database(db: SafeDBWrapper):
    """A helper to create a fresh table for the example."""
    logging.info("Setting up the 'users' table...")
    # The `with` block handles commits/rollbacks, so no need for manual calls here.
    db.execute("DROP TABLE IF EXISTS users")
    db.execute("""
        CREATE TABLE users (
            id INTEGER PRIMARY KEY,
            username TEXT NOT NULL UNIQUE,
            email TEXT
        )
    """)
    # Insert multiple records in a single transaction
    users_to_add = [
        (1, 'alice', 'alice@example.com'),
        (2, 'bob', 'bob@example.com'),
        (3, 'charlie', 'charlie@example.com'),
    ]
    for user in users_to_add:
        db.execute("INSERT INTO users VALUES (?, ?, ?)", user)
    logging.info("Database setup complete. Users inserted.")


if __name__ == "__main__":
    DB_FILE = "production_example.db"

    # Using the context manager for clean, safe handling of the connection
    try:
        with SafeDBWrapper(DB_FILE) as db:
            setup_database(db)

        print("\n--- DEMONSTRATING SAFE AND CONVENIENT FETCHING ---")
        with SafeDBWrapper(DB_FILE) as db:
            # GOOD: Fetch all users using a helper
            all_users = db.execute_and_fetch_all("SELECT id, username FROM users ORDER BY id")
            print(f"[GOOD] Fetched all users: {[dict(u) for u in all_users]}")

            # GOOD: Fetch one specific user
            user_id = 2
            user = db.execute_and_fetch_one("SELECT * FROM users WHERE id = ?", (user_id,))
            if user:
                print(f"[GOOD] Fetched user with ID {user_id}: {dict(user)}")

        print("\n--- DEMONSTRATING SQL INJECTION PREVENTION ---")
        with SafeDBWrapper(DB_FILE) as db:
            # Attacker tries to inject a payload to delete data
            malicious_input = "2; DROP TABLE users; --"

            # The parameterized query treats the input as a literal string.
            # No user has the ID "2; DROP TABLE users; --", so nothing is found.
            # The DROP TABLE command is NEVER executed.
            result = db.execute_and_fetch_one("SELECT * FROM users WHERE id = ?", (malicious_input,))
            print(f"[SAFE] Tried to fetch user with malicious ID. Result: {result}")

            # Verify that the users table still exists and is unharmed
            remaining_users = db.execute_and_fetch_all("SELECT username FROM users")
            print(f"[SAFE] Data is unharmed. Users remaining: {[u['username'] for u in remaining_users]}")

    except Exception as e:
        logging.error(f"An unexpected error occurred in main execution: {e}")
    finally:
        # Clean up the database file
        if os.path.exists(DB_FILE):
            os.remove(DB_FILE)
            logging.info(f"Cleaned up {DB_FILE}.")

